kafka-connect-ftp/src/main/scala/com/datamountaineer/streamreactor/connect/ftp/source/FtpMonitor.scala

```diff
@@ -53,77 +55,155 @@ class FtpMonitor(settings:FtpMonitorSettings, fileConverter: FileConverter) exte
+          true
+        }
+        case Some(known) if known.offset != file.ftpFile.getSize => {
+          logger.info(s"${file.name} was not completely read & committed ${known.offset} => ${file.ftpFile.getSize}")
+          true
+        }
+        case _ =>
+          logger.info(s" no fetch required")
+          false
+      }
     }
-    case _ => false
   }
 
   // Retrieves the FtpAbsoluteFile and returns a new or updated KnownFile
-  def fetch(file: AbsoluteFtpFile, prevFetch: Option[FileMetaData]): Option[(Option[FileMetaData], FetchedFile)] = {
+  def fetch(mode: MonitorMode, file: AbsoluteFtpFile, prevFetch: Option[FileMetaData]): Option[(Option[FileMetaData], FetchedFile)] = {
     logger.info(s"fetching ${file.path}")
     val baos = new ByteArrayOutputStream()
+    if( mode != MonitorMode.UpdateSlice ) {
+      if (ftp.retrieveFile(file.path, baos)) {
+        val bytes = baos.toByteArray
+        baos.close()
+        val hash = DigestUtils.sha256Hex(bytes)
+        val attributes = new FileAttributes(file.path, file.ftpFile.getSize, file.ftpFile.getTimestamp.toInstant)
+        val meta = prevFetch match {
+          case None => FileMetaData(attributes, hash, Instant.now, Instant.now, Instant.now)
+          case Some(old) => FileMetaData(attributes, hash, old.firstFetched, old.lastModified, Instant.now)
+        }
+        Option(prevFetch, FetchedFile(meta, bytes))
+      } else {
+        logger.warn(s"failed to fetch ${file.path}: ${ftp.getReplyString}")
+        None
+      }
+    } else {
+
+      val offsetToReadFrom = prevFetch match {
+        case None => 0
+        case Some(previous) if previous.offset == -1 => 0
+        case Some(previous) => previous.offset
+      }
+      logger.info(s"offsetToReadFrom= ${offsetToReadFrom}")
+
+      ftp.setRestartOffset(offsetToReadFrom)
+      val inputStream = ftp.retrieveFileStream(file.path)
+      //TODO : slice size should be set by configuration
+      var buffer = new Array[Byte](4096)
+      val bytesRead = inputStream.read(buffer)
+      inputStream.close()
+      ftp.completePendingCommand()
+
+      //Trim end
+      val bytes=util.Arrays.copyOfRange(buffer, 0, bytesRead)
+
+      logger.info(s"bytesRead= ${bytesRead}")
 
-    if (ftp.retrieveFile(file.path, baos)) {
-      val bytes = baos.toByteArray
-      baos.close()
+      // it doesn't make sense to compute the hash for chunks
       val hash = DigestUtils.sha256Hex(bytes)
-      val attributes = new FileAttributes(file.path, file.ftpFile.getSize, file.ftpFile.getTimestamp.toInstant)
+
+      val attributes = new FileAttributes(file.path, file.ftpFile.getSize , file.ftpFile.getTimestamp.toInstant)
+
       val meta = prevFetch match {
         case None => FileMetaData(attributes, hash, Instant.now, Instant.now, Instant.now)
         case Some(old) => FileMetaData(attributes, hash, old.firstFetched, old.lastModified, Instant.now)
```

Comment:
Just to be sure: FileMetaData is what's stored in kafka right ? Which means this is where we store the offset in the file ?